app/forms/ALFForms.scala

```diff
@@ -32,10 +32,25 @@ object ALFForms extends EmptyStringValidator {
 
   def messageConstants(isWelsh:Boolean): MessageConstants = if(isWelsh) WelshMessageConstants else EnglishMessageConstants
 
-  def lookupForm(isWelsh: Boolean = false) = Form(
+  def hasInvalidChars(chars: String) = !chars.replaceAll("\\s", "").forall(_.isLetterOrDigit)
+
+  def isInvalidPostcode(postcode: String) = !Postcode.cleanupPostcode(postcode).isDefined
+
+  def postcodeConstraint(isWelsh: Boolean, isUkMode: Boolean): Constraint[String] = new Constraint[String](Some("constraints.postcode"), Seq.empty)({postcode =>
+    val errors = postcode match {
+      case empty if empty.isEmpty => Seq(ValidationError(messageConstants(isWelsh).lookupPostcodeEmptyError(isUkMode)))
+      case chars if hasInvalidChars(chars) => Seq(ValidationError(messageConstants(isWelsh).lookupPostcodeInvalidError(isUkMode)))
+      case postcode if isInvalidPostcode(postcode) => Seq(ValidationError(messageConstants(isWelsh).lookupPostcodeError(isUkMode)))
+      case _ => Nil
+    }
+
+    if(errors.isEmpty) Valid else Invalid(errors)
+  })
```

Comment:
Handling this with pattern matching seems ok to me although my scala-foo is pretty limited so perhaps there's a better way to deal with it?

It's adapted from the example here https://www.playframework.com/documentation/2.8.x/ScalaCustomValidations

Reply:
I have refactored the above method a bit to make it  easier. 

I am ok with using a separate method with pattern matching to do the validation if there are more than 2 validations for a single field, which is what postcode has. It looks easier to read as well.

If we don't want to use pattern matching we can use the constraint chaining like this : [Example of constraint chaining code ](https://github.com/hmrc/ers-returns-frontend/blob/cbc1b9eccdf823266bb310ccfc9525630e7e82ee/app/models/RsFormMappings.scala#L73-L77)

Reply:
ah sure. i knew there would be another way (scala!), i guess if there's no performance impact then it's just a style preference?